Include description in repository search.

davidsvantesson opened the pull request:
When doing a keyword search for repositories (for example explore page), include repository description in search. The repository name can be something imaginary / made-up or a product number, while the description actually explains what it does. Therefore you want to include the repository description in the search. This PR should solve #7264.

Do you see any case you don't want repository description included in search, for performance issues or other reasons? It could be made a setting in app.ini, but I didn't see any reason for it.

guillep2k commented:
> 
> 
> I think it will be a performance problem on a big Gitea instance. For a smaller instance, search maybe an unnecessary things.

I imagine a big single-server system to have... 10,000 repositories? That would be about ~6MB `repository` table, most likely completely cached.

davidsvantesson commented:
@lunny Can it be made an app.ini setting (default off)? That way today behavior remains when upgrading, for those that don't want it (due to performance or search including too much).

davidsvantesson commented:
Based on comments, I have done the following:
- Added app.ini setting (SEARCH_REPO_DESCRIPTION = false)
- Added option to API whether to include description not not (default false, not affected by app.ini)
- If enabled, description will be searched on all pages where description is seen on result page: Explore page, Organization and User profile page. Also on admin repo search page (for consistency with the other search pages).
- Description not searched on "incremental" search on dashboard as you only see the repo name, also ajax makes it performance critical (it uses the API).

I also added a small test.

sapk commented on models/repo_list_test.go:
```diff
@@ -74,6 +74,34 @@ func TestSearchRepositoryByName(t *testing.T) {
 	assert.Equal(t, int64(0), count)
 
+	// Test search within description
+	repos, count, err = SearchRepository(&SearchRepoOptions{
```
Please update the test function name to reflect the new name `func TestSearchRepository(t *testing.T)` (even if it still SearchRepositoryByName the real func tested is SearchRepository)

davidsvantesson replied:
Fixed

guillep2k commented on custom/conf/app.ini.sample:
```diff
@@ -116,6 +116,8 @@ DEFAULT_THEME = gitea
 ; Whether the full name of the users should be shown where possible. If the full name isn't set, the username will be used.
 DEFAULT_SHOW_FULL_NAME = false
+; Whether to search within description at repository search on explore page.
+SEARCH_REPO_DESCRIPTION = false
```
I know that I was not the one worried about performance in this case, but I think `SEARCH_REPO_DESCRIPTION` should default to true.  @sapk, @lunny

lafriks replied:
I also agree that this should default to true

sapk replied:
ok for me

davidsvantesson replied:
I set it to false to not change behavior for existing installations, but I think most will want to have it on. @lunny Your opinion on this?

lafriks replied:
It does change behavior but we tend to enable new features by default if only they have dramatic change in performance/disk usage. I don't think this have such big impact on performance

davidsvantesson replied:
Changed to default true!

lunny commented:
> @lunny Can it be made an app.ini setting (default off)? That way today behavior remains when upgrading, for those that don't want it (due to performance or search including too much).

I think this is a good idea.

lafriks commented:
@lunny need you approval ;)
